core/task/manager.go

```diff
@@ -491,93 +490,107 @@ func (m *Manager) acquireTasks(envId uid.ID, taskDescriptors Descriptors) (err e
-
-		roOutcome := <-m.resourceOffersDone
-
-		utils.TimeTrack(timeDeployMu, "acquireTasks: deployment critical section",
-			log.WithField("tasksToRun", len(tasksToRun)).
-				WithField("partition", envId))
-
-		m.deployMu.Unlock()
-
-		deployedTasks = roOutcome.deployed
-		undeployedDescriptors = roOutcome.undeployed
-		undeployableDescriptors = roOutcome.undeployable
-
-		log.WithField("tasks", deployedTasks).
-			WithField("partition", envId).
-			Debugf("resourceOffers is done, %d new tasks running", len(deployedTasks))
-
-		if len(deployedTasks) != len(tasksToRun) {
-			// ↑ Not all roles could be deployed. If some were critical,
-			//   we cannot proceed with running this environment. Either way,
-			//   we keep the roles running since they might be useful in the future.
 			log.WithField("partition", envId).
-				Errorf("environment deployment failure: %d tasks requested for deployment, but %d deployed", len(tasksToRun), len(deployedTasks))
-
-			for _, desc := range undeployedDescriptors {
-				if desc.TaskRole.GetTaskTraits().Critical == true {
-					deploymentSuccess = false
-					undeployedCriticalDescriptors = append(undeployedCriticalDescriptors, desc)
-					printname := fmt.Sprintf("%s->%s", desc.TaskRole.GetPath(), desc.TaskClassName)
-					log.WithField("partition", envId).
-						Errorf("critical task deployment failure: %s", printname)
-				} else {
-					undeployedNonCriticalDescriptors = append(undeployedNonCriticalDescriptors, desc)
-					printname := fmt.Sprintf("%s->%s", desc.TaskRole.GetPath(), desc.TaskClassName)
-					log.WithField("partition", envId).
-						Warnf("non-critical task deployment failure: %s", printname)
+				Debugf("scheduler has been sent request to deploy %d tasks", len(tasksToRun))
+
+			timeReviveOffers := time.Now()
+			timeDeployMu := time.Now()
+			m.reviveOffersTrg <- struct{}{} // signal scheduler to revive offers
+			<-m.reviveOffersTrg             // we only continue when it's done
+			utils.TimeTrack(timeReviveOffers, "acquireTasks: revive offers",
+				log.WithField("tasksToRun", len(tasksToRun)).
+					WithField("partition", envId))
+
+			roOutcome := <-outcomeCh // blocks until a verdict from resourceOffers comes in
+
+			utils.TimeTrack(timeDeployMu, "acquireTasks: deployment critical section",
+				log.WithField("tasksToRun", len(tasksToRun)).
+					WithField("partition", envId))
+
+			deployedTasks = roOutcome.deployed
+			undeployedDescriptors = roOutcome.undeployed
+			undeployableDescriptors = roOutcome.undeployable
+
+			log.WithField("tasks", deployedTasks).
+				WithField("partition", envId).
+				Debugf("resourceOffers is done, %d new tasks running", len(deployedTasks))
+
+			if len(deployedTasks) != len(tasksToRun) {
+				// ↑ Not all roles could be deployed. If some were critical,
+				//   we cannot proceed with running this environment. Either way,
+				//   we keep the roles running since they might be useful in the future.
```

Comment:
I don't understand. Where is the part of code which tries to re-deploy only the undeployed tasks?

Reply:
I guess the comment is misleading. They would be useful in the future if task reuse was enabled, but it was disabled some years ago since start-stop-start wasn't there yet. In case of partial deployment we'd actually fail, but to my knowledge the current undeployable cases we have are actually early failures where no tasks are deployed.